Set time

peterbarker opened the pull request:
This is based on @patrickelectric 's work #7716

This is attempting to address two points:
 - tridge didn't like the #ifdefs for the different platforms we support.  Sadly, most pretend to do the `clock_gettime()` thing, so pushing this to the seperate HALs  has multiplied the code a bit.
 - Using ChibiOS time functions isn't a great move.  I grabbed @night-ghost's code from F4Light which just keeps a delta between the epoch and boot times.

peterbarker commented:
@patrickelectric Is it OK if I squash away your commits in this PR?

patrickelectric commented:
@peterbarker np.

OXINARF commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,125 @@
+        gcs().send_text(MAV_SEVERITY_WARNING, "RTC: Invalid UTC time (%u)", rtc_shift);
+        return;
+    }
+    rtc_shift = time_utc_usec - AP_HAL::micros64();
```
I wonder if this shift should be given by the caller, which should have it more accurate.

OXINARF commented on libraries/AP_HAL_Linux/Util.cpp:
```diff
@@ -91,16 +91,12 @@ void Util::_toneAlarm_timer_tick() {
-    ts.tv_sec = time_utc_usec/1000000ULL;
-    ts.tv_nsec = (time_utc_usec % 1000000ULL) * 1000ULL;
-    clock_settime(CLOCK_REALTIME, &ts);
+    return;
```
This seems totally wrong, it means all Linux boards except Linux_None won't set the time on hardware. I also don't see why Linux_None wouldn't set the time, it's still Linux, just no specific board.

OXINARF commented on ArduCopter/commands.cpp:
```diff
@@ -149,25 +149,3 @@ bool Copter::far_from_EKF_origin(const Location& loc)
-        // update signing timestamp
-        GCS_MAVLINK::update_signing_timestamp(gps_timestamp);
-
-        Log_Write_Event(DATA_SYSTEM_TIME_SET);
```
Please remove this event macro too.

OXINARF commented on ArduSub/commands.cpp:
```diff
@@ -144,25 +144,3 @@ bool Sub::far_from_EKF_origin(const Location& loc)
-        // update signing timestamp
-        GCS_MAVLINK::update_signing_timestamp(gps_timestamp);
-
-        Log_Write_Event(DATA_SYSTEM_TIME_SET);
```
Same.

peterbarker commented:
On Thu, 12 Apr 2018, Francisco Ferreira wrote:

> Also, is there any reason not to make AP_RTC a real singleton? Constructor
> has no arguments and there are no parameters so I see no reason why it
> needs to be declared in all vehicles - creating and returning an instance
> seems possible.

That's an interesting thought.  There's not a huge amount of precedence... 
just AP_HAL?  OTOH, I can't think there could never be cause to have 
parameters to the constructor... on balance I think I like the idea!

> In any case, current implementation is broken, in the sense that you can
> create multiple AP_RTC objects.

Ah, missing protection...

> In libraries/AP_RTC/AP_RTC.cpp:
> 
> > +    if (!_is_set) {
> +        return 0U;
> +    }
> +    return rtc_shift + AP_HAL::micros64();
> +}
> +
> +void AP_RTC::set(uint64_t time_utc_usec) {
> +    if (_is_set) {
> +        // you can only set once
> +        return;
> +    }
> +    if (time_utc_usec < AP_HAL::micros64()) {
> +        gcs().send_text(MAV_SEVERITY_WARNING, "RTC: Invalid UTC time (%u)
> ", rtc_shift);
> +        return;
> +    }
> +    rtc_shift = time_utc_usec - AP_HAL::micros64();
> 
> I wonder if this shift should be given by the caller, which should have it
> more accurate.

I don't think so - this is simpler.  I'm not sure what you mean by "more 
accurate".  Convince me!  I'm planning on changing this interface anyway 
to facilitate clock skewing...


>  #if CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_NONE
> -    _system_time_was_set = true;
> -
> -    timespec ts;
> -    ts.tv_sec = time_utc_usec/1000000ULL;
> -    ts.tv_nsec = (time_utc_usec % 1000000ULL) * 1000ULL;
> -    clock_settime(CLOCK_REALTIME, &ts);
> +    return;
> 
> This seems totally wrong, it means all Linux boards except Linux_None
> won't set the time on hardware. I also don't see why Linux_None wouldn't
> set the time, it's still Linux, just no specific board.

It was like that when I got here :-P  This isn't a change I need to make 
- 'though I'm hoping we can do better in the future in terms of using the 
hw clocks.  *One* reason not to set it is so people's laptops don't get 
their clocks reset - remember that we're forced to run examples with 
elevated privs ATM?


> -        Log_Write_Event(DATA_SYSTEM_TIME_SET);
> 
> Please remove this event macro too.

I'm wondering if we should make events an all-vehicle thing at some 
stage... in the meantime I'll remove it.


Thanks for looking at this.  I plan on adding clock-slew today - a 
PixRacer here slews about 0.3 seconds/hour, which is too much to let go. 
Since the way the library is called has to change (must be called with 
ever reading, no more is_set()) it's better to do it now.

rmackay9 commented:
I think @peterbarker has one more thing he wants to add to this PR related to the selection of time sources.  Then we can ask Francisco and Tridge for their (hopefully) final review.

peterbarker commented:
Primitive clock-slewing is now implemented.

When synchronising against my laptop it stays within ~7ms

Work that needs to be done on `AP_RTC`, but hopefully not as part of the initial merge:
 - determination of latency of a clock source, adjusting delta based on that
 - running a "master clock offset" rather than choosing a "blessed" time source and sticking with it.  This will involve rejection of UTC offsets which are clearly bogus, and slewing the master clock offset according to a selected backend (or selection of backends).  In particular, having a "health" set for a clock source is important; if you lose your GPS then we will continue to use the current offset so your RTC will drift.
 - dataflash logging
 - better over-mavlink debug
 - options to source from a "canonically correct" HW RTC (e.g. an ntp-synced Linux HW RTC)
 - better options to set the HW RTC (periodically?  on reboot? etc)
 - possible proper frontend/backend split
 - lock onto a single systemid/componentid for time-keeping via `SYSTEM_TIME`
 - better time-slewing algorithm
 - consider using it for other places in the system
   - GPS time is currently hard-used in places like camera logging.  This is going to be critically important for some people coordinating with external systems (GPS being a universal time all system can agree on).  It does, however, mean people taking photos indoors will have problems where gps-epoch-time is unavailable.

rmackay9 commented:
As mentioned to @peterbarker directly, I think we should avoid sending clock information to the ground station by default because it will just confused 99.9% of the users and won't add much value to them.

rmackay9 commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,239 @@
+        gcs().send_text(MAV_SEVERITY_WARNING, "RTC: too many clock sources");
+        return _invalid_source;
+    }
+    gcs().send_text(MAV_SEVERITY_WARNING, "RTC: new %s clock source",
```
I don't think we need this new clock source do we?  All is well so I don't think we should report to the GCS.

rmackay9 commented on libraries/AP_RTC/AP_RTC.h:
```diff
@@ -0,0 +1,90 @@
+
+      non-const as it updates the RTC shift as part of retrieving the value
+     */
+    uint64_t get();
```
I would probably rename this method to "get_system_time_usec" or something like that but it's not critical.

tridge replied:
I'd like to see it renamed too. I'm also dubious about it being non-const and triggering updates. Peter said its because I suggested it, but I don't remember that

rmackay9 commented on ArduCopter/Copter.h:
```diff
@@ -303,37 +303,39 @@ class Copter : public AP_HAL::HAL::Callbacks {
-            uint8_t initialised_params      : 1; // 25      // true when the all parameters have been initialised. we cannot send parameters to the GCS until this is done
-            uint8_t compass_init_location   : 1; // 26      // true when the compass's initial location has been set
-            uint8_t rc_override_enable      : 1; // 27      // aux switch rc_override is allowed
+            bool unused1                 : 1;
```
snuck in an unrelated formatting change but I don't mind really.

rmackay9 commented:
LGTM, perhaps I'd like the new clock source message to the GCS to be removed because users won't know what to do with that and nothing is wrong..  but in any case, i'm happy for this to go in.

rmackay9 commented:
@OXINARF, @tridge, whenever you guys have time..!

patrickelectric commented:
@peterbarker The idea is to send the time from GCS periodically and not only once ?

peterbarker commented:
On Tue, 17 Apr 2018, Patrick José Pereira wrote:

> @peterbarker The idea is to send the time from GCS periodically and not only once ?

Correct.

peterbarker commented:
On Tue, 17 Apr 2018, Francisco Ferreira wrote:

> There is no precedence for this in ArduPilot AFAIK, but it's the usual way to do singletons in C++.
> Now that it is in BoardConfig it's better, although I would still go with the singleton (not depending on BoardConfig to have RTC).

Yep, this came up on the dev call.  When we want to hang parameters in 
here it will be advantageous to be in the boardconfig object.

>       It was like that when I got here :-P This isn't a change I need to make - 'though I'm hoping we can do better in the future in terms of using
>       the hw clocks.
> 
> No, it wasn't... It wasn't setting the clock for Linux None board and was setting for all other boards. Now, it is doing the opposite, setting for Linux
> None and not setting for any other board.
> This bug is still present.

Oh, geeze.  That's embarassing.  Fixed.

> Now for comments on last version of this PR. Besides inline comments:
>
>  *  how did you figure that template system wasn't working? And where was it not working, SITL or fmuv4? Also, I don't get the comment that the issue
>     might be that int is typedef to int64, how would that break things?

I couldn't have something that worked in both SITL and when compiling for 
fmuv4; I'd get a redefinition when compiling for one and a 
lack-of-function-present if I removed it.  I can recreate if you like - I 
just decided it wasn't my fight and moved on.

>  *  use the new Co-author stuff from GitHub: https://help.github.com/articles/creating-a-commit-with-multiple-authors/

Thanks!

>  *  SOURCE_HW isn't registered from anywhere, which means it's unused; this breaks previous boards that could have hardware RTC set by other
>     sources/programs other than ArduPilot (get_hw_rtc is effectively unused!)

I've added a clock source to set it at boot time.

>  *  why isn't hardware RTC set when rtc_shift is changed? This would mean that ArduPilot would be changing time but any other program in the same (POSIX)
>     board would not get it

I'd prefer to have parameters before doing this.  We wouldn't want to do 
this too often unless the user specifically requested it...

> Do we really need to send this by default? Leaving this for debug is fine, but I don't think this is useful information to send continuously in release
> builds.

Agreed; will make this dependant on DEBUG

> +    }
> +    source_info &info = _source_info[source_id];
> +
> +    const int64_t delta = time_utc_usec - AP_HAL::micros64();
> +    if (info.is_set) {
> +        info.reading_average += (delta-info.reading_average)/8;
> 
> Can you explain why this divides by 8? I'm not getting that.

This is a simple smoothing thing.

> +        if (now - info.last_reading_ms > 20000) {
> +            // no reading in too long.  Something is wrong...
> 
> This seems too strong of a check, returning 0 just because no update was done recently... I think the best thing to do here would be to use the better
> source by getting the closest last_reading_ms to now.

Seems reasonable.  OTOH, "0" makes it clear that something is wrong, which 
is why I went that way.  Changing it.


> +    }
> +    gcs().send_text(MAV_SEVERITY_WARNING, "RTC: new %s clock source",
> 
> I agree with @rmackay9. Leaving it for debug is fine, but no need to send it in regular builds.

Yep.

> +    // update() should be called as fast as possible.  It slews the
> +    // RTC offset according to data received.
> +    void update();
> 
> Left-over?

Yes.

> +    source_id_t _invalid_source = (source_id_t)-1;
> This can be static const.

True.

> -    return ((long long)((ts.tv_sec * 1000) + (ts.tv_usec / 1000)));
> -#elif CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
> -    return ST2MS(chVTGetSystemTime());
> -#else
> +    return ((long long)((ts.tv_sec * 1000000) + ts.tv_usec));
> 
> Don't we need to use ULL here? Probably safer, at least. The long long isn't doing anything there either, casting to it right before returning an
> uint64_t...

Sure.  Who really knows what happens on MacOSX, 'though - and this was 
pre-existing, right?

>      if (gps_updated) {
>          // set system time if necessary
> 
> Comment left.

Will fix.


> -            uint8_t compass_init_location   : 1; // 26      // true when the compass's initial location has been set
> -            uint8_t rc_override_enable      : 1; // 27      // aux switch rc_override is allowed
> +            bool unused1                 : 1;
> 
> Why keep the unused?

Will remove.  All of these changes do mean that a specific version of the 
code will need to be looked up for a specific logfile.

> +            bool system_time_set         : 1; // true if the system time has been set from the GPS
> 
> Shouldn't this be gone?

Will remove

>          // set system time if necessary
> Comment left.

Will remove.

> +    _rtc_source = AP::rtc().register_rtc_source(AP_RTC::SOURCE_GPS);
> 
> One source per MAVLink but not one source per GPS? I think it's easier to make the MAVLink sources not send SYSTEM_TIME messages than getting GPSes to
> completely agree on time...

There's a lot of work to do to make this work for everyone.  My concern is 
that if we try to do that initially then we'll completely miss the mark.

At least notionally GPSs should be working from a common time source!

> +            // AP_HAL here
> +            return;
> +        }
> +        _singleton = this;
> +    }
> +
> +    typedef uint8_t source_id_t;
> +    typedef enum : uint8_t {
> 
> Why typedef enum? Doing enum source_type [...] should work...

../../libraries/AP_RTC/AP_RTC.h:60:9: error: ‘source_type’ does not name a 
type; did you mean ‘source_info’?

I've force-pushed this up, and marked it WIP as I need to test it on 
hardware which I don't have with me ATM.

tridge commented:
has mavlink2 signing been tested with this PR ?

peterbarker commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,240 @@
+        }
+        if (now - info.last_reading_ms > 20000) {
+            // no reading in too long.  Something is wrong...
+            if (info.last_reading_ms >= ret_last_reading_ms) {
```
Will fix.

peterbarker commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,240 @@
+            // we will use this reading anyway as something is better
+            // than nothing
+        }
+        if (info.type >= ret_source_type) {
```
Hardware sources may bare no resemblance to reality.

The mavlink ones likewise, but they're more likely to (good battery-backed RTCs), and we probably don't care - if we're trying to do machine-vision stuff we want to be synced against the camera's timestamp, and not the GPS one.

Will fix the bug; we now just bypass the "better" checks if `ret == 0`

I can't emphasise enough how much more there is to do in an RTC class....

peterbarker commented:
@OXINARF I've pushed up patches to address most of your points.

What I haven't attacked is the multiple-backends-for-GPS-library.  I do intend to support that at some stage, but it *does* complicate matters, particularly as backends get instantiated/deinstantiated at a whim.

The only usecase I set out to solve here was people trying to sync their clocks for the purposes of visual syncing.  I fully intend to come back and flesh this out - but I'm really worried that if I flesh it out too much at this early stage that when it hits user hands it won't be right and will be too inflexible to morph easily.

peterbarker commented:
@tridge I haven't managed to find any breakage with mavlink2 signing.  I think the protections about time going backwards in the signing stuff itself is too tight for that  to happen.  Maybe I'm not trying hard enough.

patrickelectric commented:
@peterbarker what is the motive to set the time back to 2000 (year) after losing contact with mavlink ? The best thing is to continue the time update, and not setting the time backward..

peterbarker commented:
@patrickelectric You're looking at the case where we lose contact with the GCS for 20 seconds without another time source?  The mavlink time source shouldn't go away, so we should still use it - unless I've screwed up.  This line here is supposed to do that: 
https://github.com/ArduPilot/ardupilot/pull/8121/files#diff-b93732e371ef7abd877dd2b60bbd3233R31

Your clock will start drifting, of course.  About 1 in 10,000 on two of my fmuv4 devices.

patrickelectric commented:
@peterbarker after losing contact with mavlink, it indeed return to 2000(year). After 20s.
https://paste.kde.org/pmqj9p4oz

peterbarker commented:
@patrickelectric OK, my guess is that it's reverted back to using the hardware-clock's time estimate.  Clearly this is unacceptable; time should never go backwards.

There's actually two problems; we should be periodically setting the hardware clock's time.

The second is that we're really going to have to have a master clock - something I was trying to avoid doing at this point as it seemed an extra layer of complication.  So in this example when you lose set-time it would start to slew towards the hardware-clock instance - since that never magically adjusts from the hardware (at the moment....) you should continue into the future without slewing.

rmackay9 commented:
we really need to get this in!!

peterbarker commented:
On Sat, 5 May 2018, Randy Mackay wrote:

> we really need to get this in!!

I now have patches that at least keep time with the GPS better.

Problem is that GPS measurements come in at 200ms intervals and our main 
loop interval is 20ms.  But not exactly, so the two intervals slide over 
one another in time, leading to beats in the timing measurements.

Most of the time a GPS packet will come in after we've done GPS processing 
for a loop, so it sits in the uart buffer while we're in 
ins.wait_for_sample().  When we kick out of there we process that packet. 
As time progresses, the windows slide over one another, and suddenly the 
packet is available to the main loop instantly (or, if you're drifting the 
other way, not available instantly any more), and you see a 20ms shift in 
the delta between your RTC clock and the GPS time.  This bounces up and 
down for a while before setting in permanently.

My current strategy is to correct any error between our current rtc shift 
and the required RTC shift over a time period equal to the sample rate for 
the GPS.  That's a pretty arbitrary way of doing things.

Running a low-pass filter doesn't really help, as these beats only happen 
about every three minutes.  A strict average of the number of microseconds 
needing adjustment over the board uptime still sees >10% variation after 
an hour of uptime.

Ideally we need to parse the GPS messages as soon as we get them from the 
GPS.  tridge is hoping to provide a simple threading API shortly.  That 
should allow us to parse the message as soon as the uart read is complete, 
and hopefully be able to pass timestamped GPS messages back to the 
frontend (main) thread.

rmackay9 commented:
@peterbarker, sounds good.  I think if the beating pattern isn't going to break any current features (which I think it can't because we don't use the 'wall clock' time for anything) we can live with the 20ms of jitter.  I'd much prefer if we can get this in and then improve it.. I'm sure you do too actually.

rmackay9 commented:
@patrickelectric, could you retest this for us to see if it resolves the problems you were seeing before?

rmackay9 commented:
one minor thing is the, "AP_GPS: stuff" commit should be renamed to be more descriptive

SubMishMar commented:
Hi @peterbarker and @rmackay9 . I git cloned this pull request from @peterbarker 's forked ardupilot directory by choosing the set-time branch to load this firmware in my skyviper drone. I waf configured the board for skyviper and after building the code successfully I uploaded it to the skyviper's FCU. But I dont see any change when I compare the results with the master branch of arcupilot. I send vision position estimate via /mavros/vision_pose/pose and I also find VISP in data log with relevant data but I dont see any output on /mavros/local_position/pose. Can anyone tell me what is going wrong here?

rmackay9 commented:
@SubMishMar, @peterbarker, I think the issue is that we need to convert the "usec" field in the vision-position-estimate messages into a local system time.  So perhaps we need a new function to do that.

patrickelectric commented:
@rmackay9 @peterbarker I'll try to take a look soon.

rmackay9 commented:
I'm not a huge fan of how the get() function updates the state.  Personally I'd prefer an independently run update() function but this is certainly not a blocker for it going in from my point of view.

rmackay9 commented:
@peterbarker, I've got [two commits here](https://github.com/rmackay9/rmackay9-ardupilot/commits/peterbarker-rtc) that maybe resolve the issue re the vision-position-estimate usec field not being converted to a system time.  I doubt if you will like them as-is but perhaps it's a starting point to resolving this important issue.

tridge commented on libraries/AP_RTC/AP_RTC.h:
```diff
@@ -0,0 +1,79 @@
+     */
+    void get_system_clock_utc(int32_t &hour, int32_t &min, int32_t &sec, int32_t &ms);
+
+    uint32_t get_time_utc(int32_t hour, int32_t min, int32_t sec, int32_t ms);
```
what does get_time_utc() do? What are its units?

tridge commented on libraries/AP_RTC/AP_RTC.h:
```diff
@@ -0,0 +1,79 @@
+        return _singleton;
+    }
+
+    static const uint8_t _max_backends = 6; // at least one for each mavlink channel
```
why are these all public? Maybe use a friend class?

tridge commented on libraries/AP_GPS/GPS_Backend.h:
```diff
@@ -28,7 +29,10 @@ class AP_GPS_Backend
     // we declare a virtual destructor so that GPS drivers can
     // override with a custom destructor if need be.
-    virtual ~AP_GPS_Backend(void) {}
+    virtual ~AP_GPS_Backend(void) {
```
are all GPS backends calling the parent destructor correctly in their own destructors?

tridge commented on libraries/AP_BoardConfig/AP_BoardConfig.cpp:
```diff
@@ -225,6 +226,11 @@ void AP_BoardConfig::init()
     hal.util->set_imu_target_temp((int8_t *)&_imu_target_temperature);
 #endif
+
+    _rtc_source = rtc.register_rtc_source(AP_RTC::SOURCE_HW);
```
this seems to be a huge change in behaviour. With this change it is now trusting the system clock for signing and for reporting the system time, with no way to turn this off. Previously both would be GPS based. The vast majority of boards don't have a usable system clock at boot, so we're just pushing garbage into AP_RTC at boot.

rmackay9 commented:
For those people waiting on this PR so we can support the vision-position-estimate messages, Peter and Tridge have come up with a new approach that does not require the flight controller and companion computer clocks to be syncronised. If you're interested in trying it, the PR from Tridge and Peter is here: ArduPilot/ardupilot#8388

patrickelectric commented:
Hi @peterbarker and @rmackay9, I tested the branch today but it's not working.
```
< SYSTEM_TIME {time_unix_usec : 946684936952167, time_boot_ms : 136928}
< SYSTEM_TIME {time_unix_usec : 946684937193001, time_boot_ms : 137169}
> SYSTEM_TIME {time_unix_usec : 998571542089453, time_boot_ms : 0}
< SYSTEM_TIME {time_unix_usec : 946684937433673, time_boot_ms : 137410}
< SYSTEM_TIME {time_unix_usec : 946684937674022, time_boot_ms : 137650}
< SYSTEM_TIME {time_unix_usec : 946684937914652, time_boot_ms : 137891}
```
As you can see, I tried to set the time with `998571542089453`, but did not work.

rmackay9 commented:
@patrickelectric, can you clarify what was the original purpose of your PR?  I had assumed that it was to help with the processing of vision-position-estimate messages but I suspect that the real reason you wanted to be able to update the system clock on the flight controller was to make the dataflash logs appear with sensible dates on them.  I think it's good if when submitting a PR to put a clear description of *why* the PR is important.  I.e. not just *what* is being changed but *why*.

peterbarker commented:
On Mon, 14 May 2018, Patrick José Pereira wrote:

> As you can see, I tried to set the time with 998571542089453, but did not
> work.

That's not a valid epoch time?

patrickelectric commented:
> By the way can you clarify what was the original purpose of your PR?
@rmackay9 
Was to fix a Sub issue, like described [here](https://github.com/ArduPilot/ardupilot/pull/7716#issue-169150669) in the original PR post. The idea is only to set the system time for our submarines, this was also described in a series of comments around the PR also to clarify why this is important for our users. 

@peterbarker 
> That's not a valid epoch time?

The board was in year 2000, this value is to sent the year 2001.

peterbarker commented:
On Tue, 15 May 2018, Patrick José Pereira wrote:

>       That's not a valid epoch time?
> 
> The board was in year 2000, this value is to sent the year 2001.

... which the code is saying it too far in the past to be valid.

patrickelectric commented:
This boards does not have RTC to track the time, they will start in Y2K.

jasonbeach commented:
Hi, I'm not sure if this is the right place to ask this or not, but I was looking at the response for the timesync message, and it appears that when something like a companion computer sends a timesync message (with only ts1 filled out) the autopilot responds with a timesync message that sets tc1 to time since boot, not unix system time (which presumably is based off GPS time?).  This means if a companion computer was to use that for time-syncing, the companion computers time would go to time since boot.  

I assume that's because we may or may not have GPS lock, but it seems we  could somehow indicate that in the response packet by setting both fields to zero, etc.  Are there other motivations for not sending GPS/unix time back in the timesync message?

tridge commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,175 @@
+    // @Param: _TYPES
+    // @DisplayName: Allowed sources of RTC time
+    // @Description: Specifies which sources of UTC time will be accepted
+    // @Bitmask: 0:GPS,1:SYSTEM_TIME,2:HW
```
the doc string should be MAVLINK not "SYSTEM_TIME".

tridge commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,175 @@
+    if (rtc_source_type == SOURCE_NONE) {
+        return false;
+    }
+    return AP_HAL::micros64() + rtc_shift;
```
need to return bool

tridge commented:
I think this is ready to merge now

patrickelectric commented:
@peterbarker did was tested in px4-v2 ? I was not able to set time.
```
> SYSTEM_TIME {time_unix_usec : 998571542089453, time_boot_ms : 0}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 84992}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 85232}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 85473}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 85714}
```
```
> SYSTEM_TIME {time_unix_usec : 53, time_boot_ms : 0}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 105215}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 105455}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 105696}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 105936}
```
```
> SYSTEM_TIME {time_unix_usec : 2, time_boot_ms : 0}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 271624}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 271864}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 272105}
< SYSTEM_TIME {time_unix_usec : 0, time_boot_ms : 272346}
```
Can you share a script or anything to allow the test of this PR ? 
I'm using a simple python script here with `system_time_send`.

patrickelectric commented on ArduCopter/Copter.h:
```diff
@@ -323,7 +323,7 @@ class Copter : public AP_HAL::HAL::Callbacks {
             uint8_t land_complete_maybe     : 1; // 14      // true if we may have landed (less strict version of land_complete)
             uint8_t throttle_zero           : 1; // 15      // true if the throttle stick is at zero, debounced, determines if pilot intends shut-down when not using motor interlock
-            uint8_t system_time_set         : 1; // 16      // true if the system time has been set from the GPS
+            uint8_t system_time_set_unused  : 1; // 16      // true if the system time has been set from the GPS
```
Is this used ?

tridge replied:
no, and its kept to keep the bit numbering

patrickelectric commented:
Also, by default the board with unix time as zero without any update. The Log in QGC appears as unknown date.
![screenshot from 2018-06-14 10-08-12](https://user-images.githubusercontent.com/1215497/41414153-972b54ec-6fbb-11e8-8edc-994b7a8fe1c2.png)

tridge commented:
@patrickelectric BRD_RTC_TYPES=1 means GPS only

patrickelectric commented:
Thank you very much for the reply @tridge.
@peterbarker awesome job, everything is working perfectly !

amilcarlucas commented on libraries/AP_RTC/AP_RTC.cpp:
```diff
@@ -0,0 +1,178 @@
+void AP_RTC::set_utc_usec(uint64_t time_utc_usec, source_type type)
+{
+    if (type >= rtc_source_type) {
+        // e.g. system-time message when we've been set by the GPS
```
Maybe add a comment:
`// return early if time has already been corrected with the worse or same precision has this function call`
